doc/authoring_command_modules/authoring_commands.md

```diff
@@ -335,10 +495,84 @@ This will show VM names only within the provided resource group.
+      ...
+```
+
+If you need access to the `cli_ctx` within your validator (for instance, to make additional REST calls) the following signature can be used:
+
+```Python
+def validate_my_arg(cmd, namespace):
+   if namespace.foo:
+      # TODO: Your custom logic here
+      client=my_client_class(cmd.cli_ctx)
+      ...
+```
+
+Validators are executed after argument parsing, and thus after the native argparse-supported `type` and `action` have been applied. A CLIError raised during a validator execution will be treated as a validation error, which will result in the display of the command usage string, as opposed to the same error raised within a custom command, which will just print the error and no validation string.
+
+***Command Validators***
+The `validator` keyword applies to commands and arguments. A command can have, at most, one validator. If supplied, then *only* this validator will be executed. Any argument-level validators will be ignored. The reason to use a command validator is if the validation sequence is important.  However, the command validator can and very often is composed to individual argument level validators. You simply define the sequence in which they execute.
+
+***Argument Validators***
+An argument can be assigned, at most, one validator. However, since a command can have many arguments, this means that a command can have many argument validators. Furthermore, since an argument context may apply to many commands, this means that this argument validator can be reused across many commands. At execution time, argument validators are executed *in random order*, so you should ensure you do not have dependencies between validators. If you do, the a command validator is the appropriate route to take. It is fine to have an argument validator involve several parameters as long as they are interdependent (for example, a validator involving a vnet name and subnet name).  
+
+## Registering Flags
+
+The recommended way to register boolean properties in the CLI is using the `get_three_state_flag` arg_type.
+
+```Python
+with self.argument_context('mymod') as c:
+  c.argument('bool_prop_enabled', arg_type=get_three_state_flag())
+```
+
+This will allow the user to specify the option as a flag or as a true/false parameter and allows for maximum convenience and consistency between create and update commands.  With the above registration, the following inputs would be accepted:
+
+```
+az mymod --bool-prop-enabled       # bool_prop_enabled = TRUE
+az mymod --bool-prop-enabled true  # bool_prop_enabled = TRUE
+az mymod --bool-prop-enabled false # bool_prop_enabled = FALSE
+
+```
+
+For flags that are switches on the command itself and not persisted as properties of a resource, the flag should be registered as follows:
+
+```Python
+with self.argument_context('mymod') as c:
+  c.argument('command_switch', action='store_true')
+```
+
+In this case, it will only accept the flag form. Do not use this for boolean properties.
```

Comment:
'Do not use this for boolean properties.'
Why not? Seems like using `'store_true` for a flag makes sense and a common pattern in the CLI.

Reply:
`three_state_flag()` is the common pattern in the CLI for Boolean properties on objects. For command-level switches that change how the command behaves, `store_true` is appropriate and three-state flag doesn't make sense. But for any kind of Boolean property that is on a create and an update, three-state flag is the way to go.